src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -775,48 +795,55 @@ impl<'tcx> Tree {
-                            // Related to https://github.com/rust-lang/rust/issues/55005.
-                            && !perm.permission.is_cell()
-                            // Only trigger UB if the accessed bit is set, i.e. if the protector is actually protecting this offset. See #4579.
-                            && perm.accessed
+                        == Some(&ProtectorKind::StrongProtector)
+                        // Don't check for protector if it is a Cell (see `unsafe_cell_deallocate` in `interior_mutability.rs`).
+                        // Related to https://github.com/rust-lang/rust/issues/55005.
+                        && !perm.permission.is_cell()
+                        // Only trigger UB if the accessed bit is set, i.e. if the protector is actually protecting this offset. See #4579.
+                        && perm.accessed
                     {
                         Err(TbError {
                             conflicting_info: &node.debug_info,
                             access_cause: diagnostics::AccessCause::Dealloc,
                             alloc_id,
                             error_offset: loc_range.start,
                             error_kind: TransitionError::ProtectedDealloc,
-                            accessed_info: match prov {
-                                ProvenanceExtra::Concrete(_) =>
-                                    Some(&args.nodes.get(start_idx).unwrap().debug_info),
-                                // We don't know from where the access came during a wildcard access.
-                                ProvenanceExtra::Wildcard => None,
-                            },
+                            accessed_info: start_idx.map(|idx| &nodes.get(idx).unwrap().debug_info),
                         }
                         .build())
                     } else {
                         Ok(())
                     }
-                },
-            )?;
+                };
+            // If we have a start index we first check its subtree in traversal order.
+            // This results in us showing the error of the closest node instead of an
+            // arbitrary one.
+            if let Some(start_idx) = start_idx {
+                TreeVisitor { nodes: &mut self.nodes, loc }.traverse_this_parents_children_other(
+                    start_idx,
+                    // Visit all children, skipping none.
+                    |_| ContinueTraversal::Recurse,
+                    |args| check_strong_protector(args.idx, args.nodes, &args.loc.perms),
+                )?;
+            }
+            // Afterward we check all tags in arbitrary order, so that we also catch
+            // protectors on different subtrees.
+            // (This unnecessarily checks the tags of `start_idx`s subtree again)
+            for (idx, _) in self.nodes.iter() {
+                check_strong_protector(idx, &self.nodes, &loc.perms)?;
+            }
```

Comment:
We should definitely not to this when there's only one tree.

Maybe it's better to just traverse the other trees as well, so that we conceptually remain on the tree level.